neverbleed.c

```diff
@@ -474,120 +476,61 @@ static void get_privsep_data(const RSA *rsa, struct st_neverbleed_rsa_exdata_t *
+        size_t first_empty;
     } keys;
     neverbleed_t *nb;
-} daemon_vars = {{PTHREAD_MUTEX_INITIALIZER}};
+} daemon_vars = {{.lock = PTHREAD_MUTEX_INITIALIZER, .first_empty = SIZE_MAX}};
 
 static RSA *daemon_get_rsa(size_t key_index)
 {
-    RSA *rsa;
+    RSA *rsa = NULL;
 
     pthread_mutex_lock(&daemon_vars.keys.lock);
-    rsa = daemon_vars.keys.keys[key_index];
-    if (rsa)
-        RSA_up_ref(rsa);
+    if (key_index < daemon_vars.keys.num_slots && daemon_vars.keys.slots[key_index].pkey != NULL)
+        rsa = EVP_PKEY_get1_RSA(daemon_vars.keys.slots[key_index].pkey);
     pthread_mutex_unlock(&daemon_vars.keys.lock);
 
     return rsa;
 }
 
-/*
- *  Returns an available slot in bit array B
- *  or if not found, returns SIZE_MAX
- */
-static size_t bita_ffirst(const uint8_t *b, const size_t tot, size_t bits)
+size_t allocate_slot(void)
 {
-    if (bits >= tot)
-        return SIZE_MAX;
-
-    uint64_t w = *((uint64_t *) b);
-    /* __builtin_ffsll returns one plus the index of the least significant 1-bit, or zero if not found */
-    uint32_t r = __builtin_ffsll(w);
-    if (r)
-        return bits + r - 1; /* adjust result */
-
-    return bita_ffirst(&b[8], tot, bits + 64);
-}
+    size_t index;
 
-/*
- * bit operation helpers for the bit-array in key_slots
- */
-#define BITMASK(b) (1 << ((b) % CHAR_BIT))
-#define BITBYTE(b) ((b) / CHAR_BIT)
-#define BITSET(a, b) ((a)[BITBYTE(b)] |= BITMASK(b))
-#define BITUNSET(a, b) ((a)[BITBYTE(b)] &= ~BITMASK(b))
-#define BITBYTES(nb) ((nb + CHAR_BIT - 1) / CHAR_BIT)
-#define BITCHECK(a, b) ((a)[BITBYTE(b)] & BITMASK(b))
-
-static void adjust_slots_reserved_size(int type, struct key_slots *slots)
-{
-#define ROUND2WORD(n) (n + 64 - 1 - (n + 64 - 1) % 64)
-    if (!slots->reserved_size || (slots->size >= slots->reserved_size)) {
-        size_t size = slots->reserved_size ? ROUND2WORD((size_t)(slots->reserved_size * 0.50) + slots->reserved_size)
-                : default_reserved_size;
-#undef ROUND2WORD
-
-        switch (type) {
-        case NEVERBLEED_TYPE_RSA:
-            if ((daemon_vars.keys.keys = realloc(daemon_vars.keys.keys, sizeof(*daemon_vars.keys.keys) * size)) == NULL)
-                dief("no memory");
-            break;
-        case NEVERBLEED_TYPE_ECDSA:
-            if ((daemon_vars.keys.ecdsa_keys = realloc(daemon_vars.keys.ecdsa_keys, sizeof(*daemon_vars.keys.ecdsa_keys) * size)) == NULL)
-                dief("no memory");
-            break;
-        default:
-            dief("invalid type adjusting reserved");
-        }
-
-        uint8_t *b;
-        if ((b = realloc(slots->bita_avail, BITBYTES(size))) == NULL)
+    if ((index = daemon_vars.keys.first_empty) != SIZE_MAX) {
+        daemon_vars.keys.first_empty = daemon_vars.keys.slots[index].next_empty;
+    } else {
+        if ((daemon_vars.keys.slots = realloc(daemon_vars.keys.slots,
+                                              sizeof(daemon_vars.keys.slots[0]) * (daemon_vars.keys.num_slots + 1))) == NULL)
```

Comment:
Previously, we were doing exponential of 1.5x, but now we are not. Do we want to revive that?

Reply:
Addressed in b29c491.